### PR TITLE
PTX-1805 VPS torpedo automation

### PR DESCRIPTION
**Pull request by nareshdesh**
Currently supports 2 cases
1. enforcement=required
2. enforcement=preferred

### CODE REVIEW

File: tests/provision/provisioning_test.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	InitInstance()
+})
+
+// This test performs VolumePlacementStrategy's replica affinity  of application
+// volume
+var _ = Describe("{ReplicaAffinity}", func() {
+	It("has to schedule app and verify the replication affinity", func() {
+
+		var vpsspec string
+		vpsrules := GetVpsRules()
```
Comment (harsh-px) [inline tests/provision/provisioning_test.go]: use `camelCase` for all variables in the PR

File: tests/provision/provisioning_test.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+			Step("get nodes and set labels: "+vkey, func() {
+				lbldata = getTestLabels(vrule.GetLabels) //TODO: function argument for getting testcase labels
+				RemoveNodeLabels(lbldata)
+				lblnode := SetNodeLabels(lbldata)
+				logrus.Info("Nodes containing label", lblnode)
+				Expect(lblnode).NotTo(BeEmpty())
+				volnodes  = pvcNodeMap(vrule.GetPvcNodeLabels,lblnode)	//TODO: function argument for vol's expected nodes
+			})
+
+			Step("Rules of volume Placement: "+vkey, func () {
```
Comment (harsh-px) [inline tests/provision/provisioning_test.go]: For all `Step` blocks, start the text with lower case in the PR. e.g "rules of volume placement..."

File: tests/provision/provisioning_test.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+				//volrules := getTestRules()
```
Comment (harsh-px) [inline tests/provision/provisioning_test.go]: why are these commented?

File: tests/provision/provisioning_rules.go [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+		vpsrules[name] = d
+	} else {
+		return fmt.Errorf("vps rule: %s is already registered", name)
+	}
+
+	return nil
+}
+
+// 
+func GetVpsRules() map[string] VpsTemplate {
```
Comment (harsh-px) [inline tests/provision/provisioning_rules.go]: type `VpsTemplate` doesn't need to be public

File: tests/provision/provisioning_test.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+		var vpsspec string
+		vpsrules := GetVpsRules()
+
+		for vkey,vrule := range vpsrules {
+			var contexts []*scheduler.Context
+			var volnodes map[string]map[string][]string 
+	
+			var lbldata  [] labelDict 
+			Step("get nodes and set labels: "+vkey, func() {
+				lbldata = getTestLabels(vrule.GetLabels) //TODO: function argument for getting testcase labels
```
Comment (harsh-px) [inline tests/provision/provisioning_test.go]: why do we need this `getTestLabels` wrapper function? All it does it calls the given function. Just call it here directly.

File: tests/provision/provisioning_test.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+				RemoveNodeLabels(lbldata)
+				lblnode := SetNodeLabels(lbldata)
+				logrus.Info("Nodes containing label", lblnode)
```
Comment (harsh-px) [inline tests/provision/provisioning_test.go]: change to `logrus.Debug`.

The PR has lot of info logs that need to be downgraded to debug.

File: tests/provision/provisioning_test.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+			for _, ctx := range contexts {
+				TearDownContext(ctx, opts)
+			}
+		}
+	})
+})
+
+
+
+
```
Comment (harsh-px) [inline tests/provision/provisioning_test.go]: lot of unneeded extra space

File: tests/provision/provisioning_test.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+//Support functions
+
+
+
+
+
+
+//-- Common function
+
+//ValidateVpsRules check applied vps rules on the app
```
Comment (harsh-px) [inline tests/provision/provisioning_test.go]: nit: `ValidateVpsRules checks applied vps rules on the app`

File: tests/provision/provisioning_test.go [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+				volnodes  = pvcNodeMap(vrule.GetPvcNodeLabels,lblnode)	//TODO: function argument for vol's expected nodes
+			})
+
+			Step("Rules of volume Placement: "+vkey, func () {
+				//volrules := getTestRules()
+				//logrus.Info("Rules to check per volume", volrules)
+				vpsspec = getVpsSpec(vrule.GetSpec)
+			})
+
+			Step("==============Launch Application============== :"+vkey, func() {
```
Comment (harsh-px) [inline tests/provision/provisioning_test.go]: use consistent string formatting for Step descriptions.

File: tests/provision/provisioning_rules.go [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+}
+func init() {
+	v := &Vpscase1{"case1",true}
+	Register(v.name, v)
+}
+
+
+//#---- Case 2 ---- enforcement: preferred
+
+type Vpscase2 struct {
```
Comment (harsh-px) [inline tests/provision/provisioning_rules.go]: `Vpscase2` doesn't need to be public

File: tests/provision/provisioning_rules.go [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+      values:
+      - "SATA"`
+*/
+	return vpsspec
+}
+
+func (v * Vpscase1) CleanVps() {
+	logrus.Infof("Cleanup test case contexti for: %v", v.name)
+}
+func init() {
```
Comment (harsh-px) [inline tests/provision/provisioning_rules.go]: `init()` function should be at bottom of the file

File: tests/provision/provisioning_rules.go [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+    enforcement: preferred
+    matchExpressions:
+    - key: media_type
+      operator: In
+      values:
+      - "SATA"`*/
+	return vpsspec
+}
+
+func (v * Vpscase2) CleanVps() {
```
Comment (harsh-px) [inline tests/provision/provisioning_rules.go]: do we . need the clean function if we don't have anything to clean?

File: drivers/scheduler/k8s/specs/mysql/px-mysql-storage.yaml [hunk 13]
```diff
@@ -16,6 +16,7 @@ parameters:
   repl: "3"
   priority_io: "high"
   io_profile: "db"
+  placement_strategy: "ssd-sata-pool-placement-spread"
```
Comment (harsh-px) [inline drivers/scheduler/k8s/specs/mysql/px-mysql-storage.yaml]: don't use the term `spread` as we no longer have spread.

File: drivers/scheduler/k8s/k8s.go [hunk 14]
```diff
@@ -149,7 +150,7 @@ func (k *K8s) Init(specDir, volDriverName, nodeDriverName string) error {
 //
 func (k *K8s) RescanSpecs(specDir string) error {
 	var err error
-	logrus.Infof("Rescanning specs for %v", specDir)
+	logrus.Infof("===Rescanning specs for %v", specDir)
```
Comment (harsh-px) [inline drivers/scheduler/k8s/k8s.go]: why the extra `===`

File: drivers/scheduler/k8s/k8s.go [hunk 15]
```diff
@@ -1216,6 +1251,23 @@ func (k *K8s) WaitForDestroy(ctx *scheduler.Context, timeout time.Duration) erro
 			}
 
 			logrus.Infof("[%v] Validated destroy of Service: %v", ctx.App.Key, obj.Name)
+		} else if obj, ok := spec.(*stork_api.Rule); ok {
```
Comment (harsh-px) [inline drivers/scheduler/k8s/k8s.go]: this PR should not be adding stork Rule code

File: tests/provision/provisioning_rules.go [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+					for _,rnode := range replicas {
+						logrus.Infof("Expected Volume Node:%v Replica Node:%v", mnode, rnode)
+						if mnode == rnode {
+							found=rnode
+							break	
+						}
+					}
+					 if found == "" {
+						logrus.Errorf("Volume '%v' does not have replica on node:'%v'", appvol,mnode)
+					  	Expect(found).NotTo(BeEmpty())
```
Comment (harsh-px) [inline tests/provision/provisioning_rules.go]: you can add the description in the `NotTo(` itself and remove the logrus above and also the `if fount == ""` check.

The point of Expect is to internally check the condition. So we should not be checking for ` if found == "" {`

Comment (harsh-px) [reply tests/provision/provisioning_rules.go]: Do the same for other places in the PR.

File: drivers/scheduler/errors.go [hunk 17]
```diff
@@ -313,3 +313,15 @@ type ErrFailedToDecommissionNode struct {
...
+// ErrFailedToProvision  error type when fail to provision volume
+type ErrFailedToProvision struct {
+	// Node where the label or volume could not be applied
+	Node node.Node
+	// Cause is the underlying cause of the error
+	Cause string
+}
+
+func (e *ErrFailedToProvision) Error() string {
+	return fmt.Sprintf("Failed to label node: %v due to err: %v", e.Node, e.Cause)
```
[the remark this answers is not in the document]
Comment (nareshdesh) [reply drivers/scheduler/errors.go]: Improved the message

File: tests/provision/provisioning_rules.go [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+
+type labelDict map[string]interface{}
+
+type VpsTemplate interface {
+	// Node label
+	GetLabels()  [] labelDict
+	// Pvc label
+	GetPvcNodeLabels( lblnodes map[string][]string) map[string]map[string][]string
+	// Vps Spec
+	GetSpec ( ) string
```
[the remark this answers is not in the document]
Comment (nareshdesh) [reply tests/provision/provisioning_rules.go]: Did `go fmt ` now

File: tests/provision/provisioning_rules.go [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+
+	var vpsspec string 
+	//logrus.Infof(" rules:%v ", rules)	
+	vpsspec =`apiVersion: portworx.io/v1beta2
+kind: VolumePlacementStrategy
+metadata:
+  name: ssd-sata-pool-placement-spread
+spec:
+  replicaAffinity:
+#  - affectedReplicas: 1
```
[the remark this answers is not in the document]
Comment (nareshdesh) [reply tests/provision/provisioning_rules.go]: Removed affectedReplicas

File: tests/provision/provisioning_rules.go [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+	GetLabels()  [] labelDict
+	// Pvc label
+	GetPvcNodeLabels( lblnodes map[string][]string) map[string]map[string][]string
+	// Vps Spec
+	GetSpec ( ) string
+	// Validate 
+	Validate (appVolumes []*volume.Volume,volscheck map[string]map[string][]string)
+	// Clean up
+	CleanVps()
+	//
```
[the remark this answers is not in the document]
Comment (nareshdesh) [reply tests/provision/provisioning_rules.go]: Removed it

File: drivers/scheduler/k8s/k8s.go (hunk 14 above)
Comment (nareshdesh) [reply drivers/scheduler/k8s/k8s.go]: Removed it

File: tests/asg/asg_test.go [hunk 21]
```diff
@@ -43,7 +44,7 @@ var _ = Describe("{ClusterScaleUpDown}", func() {
 
 		var contexts []*scheduler.Context
 		for i := 0; i < Inst().ScaleFactor; i++ {
-			contexts = append(contexts, ScheduleAndValidate(fmt.Sprintf("asgscaleupdown-%d", i))...)
+			contexts = append(contexts, ScheduleAndValidate(fmt.Sprintf("asgscaleupdown-%d", i), nil,defaultVstate)...)
```
Comment (mborodin-px) [inline tests/asg/asg_test.go]: please run go fmt to format code, it should be additional space after comma

File: tests/common.go [hunk 22]
```diff
@@ -227,23 +254,29 @@ func DeleteVolumesAndWait(ctx *scheduler.Context) {
 }
 
 // ScheduleAndValidate schedules and validates applications
-func ScheduleAndValidate(testname string) []*scheduler.Context {
+func ScheduleAndValidate(testname string,vps_map map[string]string,vState int) []*scheduler.Context {
```
Comment (mborodin-px) [inline tests/common.go]: do not change ScheduleAndValidate common function just for your test, use **Inst().S.Schedule(taskName, scheduler.ScheduleOptions{...}** from your test directly with **ValidateContext(ctx,vState)**. In this case you don't need to modify all ScheduleAndValidate invocation in each tests.

**Comment by harsh-px** [comment]
No longer valid. @nareshdesh to open a new one.

**Pull request closed without merging**